presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionTreeUtils.java

```diff
@@ -112,4 +123,45 @@ public static boolean isEqualComparisonExpression(Expression expression)
     {
         return expression instanceof ComparisonExpression && ((ComparisonExpression) expression).getOperator() == ComparisonExpression.Operator.EQUAL;
     }
+
+    static Optional<EnumType> tryResolveEnumLiteralType(QualifiedName qualifiedName, TypeManager typeManager)
+    {
+        Optional<QualifiedName> prefix = qualifiedName.getPrefix();
+        if (!prefix.isPresent()) {
+            // an enum literal should be of the form `MyEnum.my_key`
+            return Optional.empty();
+        }
+        try {
+            Type baseType = typeManager.getType(parseTypeSignature(prefix.get().toString()));
+            if (baseType instanceof EnumType) {
+                return Optional.of((EnumType) baseType);
+            }
+        }
+        catch (IllegalArgumentException e) {
+            return Optional.empty();
+        }
+        return Optional.empty();
+    }
+
+    public static boolean isEnumLiteral(DereferenceExpression node, Type nodeType)
+    {
+        if (!(nodeType instanceof EnumType)) {
+            return false;
+        }
+        QualifiedName qualifiedName = DereferenceExpression.getQualifiedName(node);
+        requireNonNull(qualifiedName, "qualifiedName is null");
+        Optional<QualifiedName> prefix = qualifiedName.getPrefix();
+        return prefix.isPresent()
+                && prefix.get().toString().equalsIgnoreCase(nodeType.getTypeSignature().getBase());
+    }
+
+    public static Object resolveEnumLiteral(DereferenceExpression node, EnumType enumType)
+    {
+        QualifiedName qualifiedName = DereferenceExpression.getQualifiedName(node);
+        checkArgument(qualifiedName != null && isEnumLiteral(node, enumType), "Node is not an enum literal");
```

Comment:
This is a minor issue: Seems like you are always calling `isEnumLiteral` and `resolveEnumLiteral` together and then you can checking `isEnumLiteral` here again. So maybe worth have a `tryResolveEnumLiteral` that returns a `Optional<Object>` to avoid calling `isEnumLiteral` twice.

Reply:
ha that's what I was doing before and you told me to change it to a precondition check 😆

Reply:
Ya, i kinda feel like i probably suggested this 🤦‍♀️. You can keep the `isEnumLiteral` function. This is more to suggest to only call 1 function at the call sites rather than changing this implementation (other than changing the return type to optional).